ddtrace/context.py

```diff
@@ -1,98 +1,102 @@
-from .utils.formats import get_env
 
 
 if TYPE_CHECKING:
-    from ddtrace import Span
+    from .span import Span
+    from .span import _MetaDictType
+    from .span import _MetricDictType
 
 log = get_logger(__name__)
 
 
+@attr.s(eq=False, slots=True)
 class Context(object):
+    """Represents the state required to propagate a trace across execution
+    boundaries.
     """
-    Context is used to keep track of a hierarchy of spans for the current
-    execution flow. During each logical execution, the same ``Context`` is
-    used to represent a single logical trace, even if the trace is built
-    asynchronously.
-
-    A single code execution may use multiple ``Context`` if part of the execution
-    must not be related to the current tracing. As example, a delayed job may
-    compose a standalone trace instead of being related to the same trace that
-    generates the job itself. On the other hand, if it's part of the same
-    ``Context``, it will be related to the original trace.
-
-    This data structure is thread-safe.
-    """
-
-    _partial_flush_enabled = asbool(get_env("tracer", "partial_flush_enabled", default=False))
-    _partial_flush_min_spans = int(get_env("tracer", "partial_flush_min_spans", default=500))  # type: ignore[arg-type]
-
-    def __init__(
-        self,
-        trace_id=None,  # type: Optional[int]
-        span_id=None,  # type: Optional[int]
-        sampling_priority=None,  # type: Optional[int]
-        dd_origin=None,  # type: Optional[str]
-    ):
-        # type: (...) -> None
-        """
-        Initialize a new thread-safe ``Context``.
-
-        :param int trace_id: trace_id of parent span
-        :param int span_id: span_id of parent span
-        """
-        self._current_span = None  # type: Optional[Span]
-        self._lock = threading.RLock()
 
-        self._parent_trace_id = trace_id
-        self._parent_span_id = span_id
-        self._sampling_priority = sampling_priority
-        self._local_root_span = None  # type: Optional[Span]
-        self.dd_origin = dd_origin
+    trace_id = attr.ib(default=None, type=Optional[int])
+    span_id = attr.ib(default=None, type=Optional[int])
+    _dd_origin = attr.ib(default=None, type=Optional[str])
+    _sampling_priority = attr.ib(default=None, type=Optional[NumericType])
+    _lock = attr.ib(factory=threading.RLock, type=threading.RLock)
+    _meta = attr.ib(factory=dict)  # type: _MetaDictType
+    _metrics = attr.ib(factory=dict)  # type: _MetricDictType
+
+    def __attrs_post_init__(self):
+        self.dd_origin = self._dd_origin
+        self.sampling_priority = self._sampling_priority
+        del self._dd_origin
+        del self._sampling_priority
+
+    def __eq__(self, other):
+        with self._lock:
+            return (
+                self.trace_id == other.trace_id
+                and self.span_id == other.span_id
+                and self._meta == other._meta
+                and self._metrics == other._metrics
```

Comment:
are these really needed for equality? seems a context would be equal if the trace/span ids are equal? should that be enough?

Reply:
sampling priority and dd_origin i would think have to match for contexts to be equal but I can't think of a scenario where you'd have matching trace_id and span_id and not sampling priority or origin...

Reply:
do we ever expect the context to be split, and us needing to compare them?

if contexts are split, sampling priority can be different if the user manually sets it on one context but not the other.

for origin, I don't think it should ever (?) differ, because we don't allow users to set it manually.

Reply:
> do we ever expect the context to be split, and us needing to compare them?

I don't think so - especially for the latter part. I don't think I know of a case where contexts would be compared except for our testing (in which case meta/metrics would matter)

Reply:
ok, so this is for testing only?

then it is probably ok to leave this as-is.